Lib/pdb.py

```diff
@@ -3553,28 +3561,48 @@ def main():
         color=True,
     )
 
-    # We need to maunally get the script from args, because the first positional
-    # arguments could be either the script we need to debug, or the argument
-    # to the -m module
+    # Get all the commands out first. For backwards compatibility, we allow
+    # -c commands to be after the target.
     parser.add_argument('-c', '--command', action='append', default=[], metavar='command', dest='commands',
                         help='pdb commands to execute as if given in a .pdbrc file')
-    parser.add_argument('-m', metavar='module', dest='module')
-    parser.add_argument('-p', '--pid', type=int, help="attach to the specified PID", default=None)
 
-    if len(sys.argv) == 1:
+    opts, args = parser.parse_known_args()
+
+    if not args:
         # If no arguments were given (python -m pdb), print the whole help message.
         # Without this check, argparse would only complain about missing required arguments.
+        # We need to add the arguments definitions here to get a proper help message.
+        parser.add_argument('-m', metavar='module', dest='module')
+        parser.add_argument('-p', '--pid', type=int, help="attach to the specified PID", default=None)
         parser.print_help()
         sys.exit(2)
+    elif args[0] == '-p' or args[0] == '--pid':
```

Comment:
Is -p always the first arg?

Reply:
So we already skimmed all the `-c` commands, which are the only "valid options" to all three kinds of invocation. Then we need to determine which path (script, module or pid) we should take. There's no case where `-p` is not the first argument (after skimming `-c`) but we should still consider a pid attach.